libs/shared/flagd-core/src/lib/targeting/targeting.ts

```diff
@@ -0,0 +1,31 @@
+import {LogicEngine,} from "json-logic-engine";
+import {endsWithHandler, endsWithRule, startsWithHandler, startsWithRule} from "./string-comp";
+import {semVer, semVerRule} from "./sem-ver";
+import {fractional, fractionalRule} from "./fractional";
+import {flagdPropertyKey, flagKeyPropertyKey, timestampPropertyKey} from "./common";
+
+export class Targeting {
+  private readonly _logicEngine: LogicEngine;
+
+  constructor() {
+    const engine = new LogicEngine();
+    engine.addMethod(startsWithRule, startsWithHandler);
+    engine.addMethod(endsWithRule, endsWithHandler);
+    engine.addMethod(semVerRule, semVer);
+    engine.addMethod(fractionalRule, fractional);
+
+    this._logicEngine = engine;
+  }
+
+  applyTargeting(flagKey: string, logic: any, data: Record<any, any>): any {
+    const ctxData = {
+      [flagdPropertyKey]: {
+        [flagKeyPropertyKey]: flagKey,
+        [timestampPropertyKey]: Date.now(),
+      },
+      ...data
+    }
```

Comment:
```suggestion
      ...data,
      [flagdPropertyKey]: {
        [flagKeyPropertyKey]: flagKey,
        [timestampPropertyKey]: Date.now(),
      },
    }
```

https://flagd.dev/reference/specifications/in-process-providers/#flagd-properties-in-the-evaluation-context

Reply:
Please also add a test to make sure the `$flagd` properties override anything supplied by the user.

Reply:
Is it the ordering of properties ? if so does it matter ? 🤔

Reply:
The ordering impacts which one "wins" in cases of duplicate keys. In your original implementation, the data would overwrite the injected properties if it had `$flagd` in it. You should check do see if the prop is incoming, and warn as we do in Go that the `$flagd` prop is special and will be overwritten.

Reply:
thanks for the details. I will add the check and warning

Reply:
Done. Added the warning and validation tests - https://github.com/open-feature/js-sdk-contrib/pull/623/commits/dd9025fabd09be28e0a23c8f4edf0b83316dcba2